src/common/libjob/specutil.h

```diff
@@ -0,0 +1,119 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#ifndef _JOB_SPECUTIL_H
+#define _JOB_SPECUTIL_H
+
+#include <jansson.h>
+
```

Comment:
OK, that makes sense. What about setting arbitrary values on the jobspec, though? At the moment that can be done by the library user constructing a ``json_t`` and passing it to ``specutil_attr_set``.

Reply:
It's probably not a common use-case but I know that @ardangelo will need to set the stop-tasks-in-exec attribute.

Reply:
What we've done in similar situations is use a setter that takes JSON string, with a corresponding `pack()` style function for convenience. The first example that comes to mind is the `flux_plugin_arg_t` interface:

```C
/*  Flags for flux_plugin_arg_get/set/pack/unpack
 */
enum {
    FLUX_PLUGIN_ARG_IN =  0,    /* Operate on input args    */
    FLUX_PLUGIN_ARG_OUT = 1,    /* Operate on output args   */
    FLUX_PLUGIN_ARG_UPDATE = 2  /* Update args for set/pack */
};

/*  Get/set arguments in plugin arg object using JSON encoded strings
 */
int flux_plugin_arg_set (flux_plugin_arg_t *args,
                         int flags,
                         const char *json_str);
int flux_plugin_arg_get (flux_plugin_arg_t *args,
                         int flags,
                         char **json_str);

/*  Pack/unpack arguments into plugin arg object using jansson pack style args
 */
int flux_plugin_arg_pack (flux_plugin_arg_t *args, int flags,
                          const char *fmt, ...);
int flux_plugin_arg_vpack (flux_plugin_arg_t *args, int flags,
                           const char *fmt, va_list ap);

```

In this case, you may want to only allow "updates", which internally create a `json_t` and use that with `json_object_update()`. Not sure if we'd want separate `setopt` and `setattr` equivalents like we have in the Python Jobspec class.

E.g. for your use case of the `stop-tasks-in-exec` shell option, it might look something like:

```C
   flux_jobspecv1_setopt_pack (spec, 0, "{s:i}", "stop-tasks-in-exec", 1);
```

Reply:
Maybe a varargs "pack" interface similar to `flux_msg_pack()`?
```c
int flux_jobspec1_attr_set (flux_jobspec1_t *jobspec, const char *path, const char *fmt, ...);
```
This lets you you set values with different scalar types, e.g.
```c
flux_jobspec1_attr_set (jobspec, "system.shell.options.stop-tasks-in-exec", "i", 1);
```
It  also lets you pass in a `json_t *` using the "o" or "O" format descriptors if the API user chooses jansson, but the main thing is it allows you to pass in any scalar types or build simple objects without committing to that library as a dependency.

Reply:
Heh, sorry for the redundant info @jameshcorbett .  I guess we are eager to help!

Reply:
Yes, this is pretty important for debugging tools like @ardangelo's. 

@ardangelo and @jameshcorbett : this also begs a question. If this were done with CLI, the debugger front end will put a break point into `MPIR_Breakpoint` within the launcher, which will be called when the target application processes hit the `stop-tasks-exec` state. 

But since we will do this with Flux APIs, I suspect CTI's flux backend will orchestrate this tool-Flux initial synchronization by writing its own code perhaps similar to https://github.com/flux-framework/flux-core/blob/master/src/cmd/flux-job.c#L1782

Do we have a Issue ticket to track the discussions for this? If not, this might warrant a ticket either in flux-core or CTI repo.

Reply:
> sorry for the redundant info @jameshcorbett . I guess we are eager to help!

Yeah, apologies, but thanks for stepping up to this task!!

Reply:
Thanks for the feedback, I updated the branch to remove ``json_t`` among a few other changes.

About the ``attr_set`` and ``attr_pack`` functionality---should it only be for setting ``attributes.*`` or should it be able to set anything on the jobspec? (I.e. should ``path`` be prefixed with ``"attributes."``?)

Reply:
I'd lean towards only setting `attribiutes.*`, sort of like:
https://github.com/flux-framework/flux-core/blob/2bef19bc11c2c501a3d0dfba028c259bff211c71/src/bindings/python/flux/job/Jobspec.py#L401

Reply:
My thought was that in Python, users can (through other easily-accessible means) modify the jobspec however they like, including entries not in ``attributes.*``. In Python, ``settattr`` is just a helper method, whereas in C, the ``attr_set`` and ``attr_pack`` would be the only interface to modify the jobspec in arbitrary ways. That said, though, I doubt many users would want to modify anything other than ``attributes.*`` and if they did they might shoot themselves in the foot...

Reply:
Right, I resonate with your last point more than your first :-)

If the API support for `tasks` and `resources` is lacking in some way that forces people into an end-run, we should prefer that they open a bug so we can fix it for the next person, rather than make the end-run too easy.

Reply:
That sounds like a good approach to me, I just thought there was something to be said for allowing all kinds of modifications and wanted to mention it.